### ISSUE TITLE
Road casings ordering leads to "bridge" like rendering

### ISSUE THREAD

**Issue by Edefritz** [protomaps/basemaps#42]
Right now all road casings are rendered just before the road type they belong to. This leads to some weird effects at intersections between different road types (minor-medium/medium-major ...), where it appears that the higher class road passes above the lower class road, instead of being just a regular intersection.

<img width="700" alt="Screenshot 2023-06-12 at 11 13 25" src="https://github.com/protomaps/basemaps/assets/1841670/84e6794b-e124-4321-ba28-96e34cd623ec">
<img width="700" alt="Screenshot 2023-06-12 at 11 13 38" src="https://github.com/protomaps/basemaps/assets/1841670/a1c62745-87b6-4de3-9ba9-e9079c675f8d">

I'm not a cartography expert, but I think road casings should all be rendered before (below) any of the roads. I will propose a PR later today, if that's ok.

**Comment by nvkelso** [comment]
+1

**Comment by wipfli** [comment]
Thanks for this contribution, @Edefritz! I agree that in general it is better to do first all casings, then all line fills for roads.

Nevertheless, I would like to mention that there are also downsides to first render all casings, and then all line fills:

1) Roads of lower priority which merge perpendicularly into a road with higher priority will "bite out" a section of the casing of the high priority road which breaks the visual continuity of the high priority road. This effect is in particular visible at low zooms.
2) A low priority road parallel and close to a high priority road will delete the casing of the high priority road. It looks effectively like the high priority road has no casing. Again, this effect is strongest at low zooms.

Here is an example from the Wikipedia basemap which has these problems (https://maps.wikimedia.org/#15/47.2109/7.9993):

![Screenshot_20230625_214632](https://github.com/protomaps/basemaps/assets/53421382/11beb957-03c6-4045-a0e6-6c0ab4735d6d)

Google Maps solves this issue by using casing-fill-casing-fill order at low zooms, while switching to casing-casing-fill-fill at higher zooms. Here is an example:

At z13: 

https://www.google.com/maps/@47.2076993,8.0005265,13z?entry=ttu

![image](https://github.com/protomaps/basemaps/assets/53421382/417a9de3-b92a-4cde-9a3d-f603e672a8ac)

At z14:

https://www.google.com/maps/@47.2076993,8.0005265,14z?entry=ttu

![image](https://github.com/protomaps/basemaps/assets/53421382/a8f17c59-615c-487c-86e0-6fe77a5d86df)


### Side Note: Bridges

The visual continuity of roads in Google Maps is something which sets it apart from other basemaps in my opinion. This does not only affect casing ordering, but also the z-ordering of roads (bridges/tunnels). At low zooms, they show high priority roads on top of low priority roads, independent of the physical ordering. `highway=motorway` will always shown on top of `highway=primary` at low zooms. Only once you zoom in enough, the physical ordering is actually shown.

Example 

At z15:

https://www.google.com/maps/@47.2095108,7.9834262,15z?entry=ttu

![Screenshot_20230625_220113](https://github.com/protomaps/basemaps/assets/53421382/b2ce39e2-4579-4be4-a0c4-aaa5141c42da)

At z16:

https://www.google.com/maps/@47.2095108,7.9834262,16z?entry=ttu

![Screenshot_20230625_220126](https://github.com/protomaps/basemaps/assets/53421382/4d1ae94b-fb23-46fe-85fe-8efc5c9db60b)

**Comment by Edefritz** [comment]
Thanks for the explanation @wipfli! I wasn't aware about all the nuances when it comes to casings and suspected a simple bug.
I guess now that my MR was accepted we can close that issue as soon as a new version of the npm package is released.

**Comment by bdon** [comment]
1.3.1 on npm: https://www.npmjs.com/package/protomaps-themes-base

**Comment by Edefritz** [comment]
Thanks so much. I will close the issue then :)
